### PR TITLE
add table aws_ssm_managed_instance_compliance. Closes #323

### LINKED ISSUE #323: Add compliance status for patch and association
**Is your feature request related to a problem? Please describe.**
Trying to check below use cases, but not getting Patch status in any of the existing ssm tables such as `aws_ssm_association` and `aws_ssm_patch_baseline`. Would like to know whether we have to add any new column to these exiting tables to get the compliance status.  Or it is there but I am unable to find ?

[SSM.2] This AWS control checks whether the compliance status of the Amazon EC2 Systems Manager patch compliance is COMPLIANT or NON_COMPLIANT after the patch installation on the instance. It only checks instances that are managed by AWS Systems Manager Patch Manager
[SSM.3] This AWS control checks whether the status of the AWS Systems Manager association compliance is COMPLIANT or NON_COMPLIANT after the association is executed on an instance

**Additional context**

https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html#ssm-2-remediation
https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-standards-fsbp-controls.html#ssm-3-remediation

### PR DESCRIPTION
**Pull request by bigdatasourav**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
SETUP: tests/aws_ssm_managed_instance_compliance []

PRETEST: tests/aws_ssm_managed_instance_compliance

TEST: tests/aws_ssm_managed_instance_compliance
Running terraform
aws_default_subnet.default_subnet: Creating...
aws_default_vpc.default: Creating...
aws_iam_role.named_test_resource: Creating...
aws_default_subnet.default_subnet: Creation complete after 3s [id=subnet-0435e4ab25bb9123e]
aws_default_vpc.default: Still creating... [10s elapsed]
aws_iam_role.named_test_resource: Still creating... [10s elapsed]
aws_iam_role.named_test_resource: Creation complete after 13s [id=turbottest29203]
aws_iam_role_policy_attachment.test-attach: Creating...
aws_iam_role_policy_attachment.test-attach: Creation complete after 2s [id=turbottest29203-20210607143554282400000001]
aws_iam_instance_profile.named_test_resource: Creating...
aws_iam_instance_profile.named_test_resource: Creation complete after 4s [id=turbottest29203]
aws_instance.named_test_resource: Creating...
aws_default_vpc.default: Still creating... [20s elapsed]
aws_default_vpc.default: Creation complete after 27s [id=vpc-080335a29453b49a1]
aws_instance.named_test_resource: Still creating... [10s elapsed]
aws_instance.named_test_resource: Still creating... [20s elapsed]
aws_instance.named_test_resource: Still creating... [30s elapsed]
aws_instance.named_test_resource: Still creating... [40s elapsed]
aws_instance.named_test_resource: Still creating... [50s elapsed]
aws_instance.named_test_resource: Still creating... [1m0s elapsed]
aws_instance.named_test_resource: Still creating... [1m10s elapsed]
aws_instance.named_test_resource: Still creating... [1m20s elapsed]
aws_instance.named_test_resource: Still creating... [1m30s elapsed]
aws_instance.named_test_resource: Still creating... [1m40s elapsed]
aws_instance.named_test_resource: Creation complete after 1m48s [id=i-056468ad4d4a7f380]
null_resource.delay: Creating...
null_resource.delay: Provisioning with 'local-exec'...
null_resource.delay (local-exec): Executing: ["/bin/sh" "-c" "sleep 300"]
null_resource.delay: Still creating... [10s elapsed]
null_resource.delay: Still creating... [20s elapsed]
null_resource.delay: Still creating... [30s elapsed]
null_resource.delay: Still creating... [40s elapsed]
null_resource.delay: Still creating... [50s elapsed]
null_resource.delay: Still creating... [1m0s elapsed]
null_resource.delay: Still creating... [1m10s elapsed]
null_resource.delay: Still creating... [1m20s elapsed]
null_resource.delay: Still creating... [1m30s elapsed]
null_resource.delay: Still creating... [1m40s elapsed]
null_resource.delay: Still creating... [1m50s elapsed]
null_resource.delay: Still creating... [2m0s elapsed]
null_resource.delay: Still creating... [2m10s elapsed]
null_resource.delay: Still creating... [2m20s elapsed]
null_resource.delay: Still creating... [2m30s elapsed]
null_resource.delay: Still creating... [2m40s elapsed]
null_resource.delay: Still creating... [2m50s elapsed]
null_resource.delay: Still creating... [3m0s elapsed]
null_resource.delay: Still creating... [3m10s elapsed]
null_resource.delay: Still creating... [3m20s elapsed]
null_resource.delay: Still creating... [3m30s elapsed]
null_resource.delay: Still creating... [3m40s elapsed]
null_resource.delay: Still creating... [3m50s elapsed]
null_resource.delay: Still creating... [4m0s elapsed]
null_resource.delay: Still creating... [4m10s elapsed]
null_resource.delay: Still creating... [4m20s elapsed]
null_resource.delay: Still creating... [4m30s elapsed]
null_resource.delay: Still creating... [4m41s elapsed]
null_resource.delay: Still creating... [4m51s elapsed]
null_resource.delay: Creation complete after 5m0s [id=718339978149547338]
null_resource.list_compliance: Creating...
null_resource.list_compliance: Provisioning with 'local-exec'...
null_resource.list_compliance (local-exec): Executing: ["/bin/sh" "-c" "aws ssm list-compliance-items --resource-ids i-056468ad4d4a7f380 --resource-types ManagedInstance --output json > /private/var/folders/d8/gsp6f_cn3359xnt_gs3cts_80000gn/T/tests/aws_ssm_managed_instance_compliance/terraform/test/compliance.json"]
null_resource.list_compliance: Creation complete after 5s [id=1769456156225228283]
data.local_file.input: Reading...
data.local_file.input: Read complete after 0s [id=51174fd3caed92d3982f8a7521df2f7b502b4934]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 8 added, 0 changed, 0 destroyed.

Outputs:

arn = "arn:aws:ssm:us-east-1:013122550996:managed-instance-compliance/5e0e8560-11c0-45d0-962b-715329d5ab31"
compliance_id = "5e0e8560-11c0-45d0-962b-715329d5ab31"
resource_id = "i-056468ad4d4a7f380"
status = "COMPLIANT"

Running SQL query: test-list-query.sql
[
  {
    "id": "5e0e8560-11c0-45d0-962b-715329d5ab31",
    "resource_id": "i-056468ad4d4a7f380",
    "status": "COMPLIANT"
  }
]
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "arn:aws:ssm:us-east-1:013122550996:managed-instance-compliance/5e0e8560-11c0-45d0-962b-715329d5ab31"
    ],
    "title": "5e0e8560-11c0-45d0-962b-715329d5ab31"
  }
]
✔ PASSED

POSTTEST: tests/aws_ssm_managed_instance_compliance

TEARDOWN: tests/aws_ssm_managed_instance_compliance

SUMMARY:

1/1 passed.

```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
> select
  id,
  resource_id,
  akas
from
  aws_ssm_managed_instance_compliance where resource_id = 'i-02a49c5a11ed9d37f'
+--------------------------------------+---------------------+----------------------------------------------------------------------------------------------------------------------------------------------
| id                                   | resource_id         | akas                                                                                                                                         
+--------------------------------------+---------------------+----------------------------------------------------------------------------------------------------------------------------------------------
| perl-macros.x86_64                   | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/perl-macros.x86_64:Patch"]                         
| nss-tools.x86_64                     | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/nss-tools.x86_64:Patch"]                           
| libxml2-python.x86_64                | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/libxml2-python.x86_64:Patch"]                      
| dhcp-libs.x86_64                     | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/dhcp-libs.x86_64:Patch"]                           
| openssl-libs.x86_64                  | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/openssl-libs.x86_64:Patch"]                        
| libcrypt.x86_64                      | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/libcrypt.x86_64:Patch"]                            
| python-pillow.x86_64                 | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/python-pillow.x86_64:Patch"]                       
| bind-export-libs.x86_64              | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/bind-export-libs.x86_64:Patch"]                    
| glibc-all-langpacks.x86_64           | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/glibc-all-langpacks.x86_64:Patch"]                 
| gnupg2.x86_64                        | i-02a49c5a11ed9d37f | ["arn:aws:ssm:us-east-1:013122550996:managed-instance/i-02a49c5a11ed9d37f/compliance-item/gnupg2.x86_64:Patch"]                              


```
</details>

### CODE REVIEW

File: aws/table_aws_ssm_managed_instance_compliance.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+func getSSMManagedInstanceComplianceTitle(_ context.Context, d *transform.TransformData) (interface{}, error) {
+	data := d.HydrateItem.(*ssm.ComplianceItem)
+
+	title := *data.Id
+	if len(*data.Title) > 0 {
+		title = *data.Title
+	}
+
+	return title, nil
+}
```
Comment (Subhajit97) [inline aws/table_aws_ssm_managed_instance_compliance.go:156]: Do we need this function for `title`?
Can't we directly use `Id` as title?

File: docs/tables/aws_ssm_managed_instance_compliance.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  name,
+  resource_id,
+  status,
+  compliance_type,
+  severity
+from
+  aws_ssm_managed_instance_compliance;
+```
+
+### List non compliant associations of managed instances
```
Comment (Subhajit97) [inline docs/tables/aws_ssm_managed_instance_compliance.md:23]:
```suggestion
### List non-compliant associations of managed instances
```

File: docs/tables/aws_ssm_managed_instance_compliance.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  compliance_type,
+  severity
+from
+  aws_ssm_managed_instance_compliance
+where
+  compliance_type = 'Association'
+  and status <> 'COMPLIANT';
+```
+
+### List non compliant patches of managed instances
```
Comment (Subhajit97) [inline docs/tables/aws_ssm_managed_instance_compliance.md:40]:
```suggestion
### List non-compliant patches of managed instances
```

File: aws/table_aws_ssm_managed_instance_compliance.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+				Description: "A Key:Value combination details for the compliance item.",
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "execution_summary",
+				Description: "A summary for the compliance item.",
+				Type:        proto.ColumnType_JSON,
+			},
+
+			// Steampipe Standard columns
```
Comment (Subhajit97) [inline aws/table_aws_ssm_managed_instance_compliance.go:73]:
```suggestion
			// Steampipe standard columns
```

File: aws/table_aws_ssm_managed_instance_compliance.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "severity",
+				Description: "The severity of the compliance status.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "details",
+				Description: "A Key:Value combination details for the compliance item.",
```
Comment (Subhajit97) [inline aws/table_aws_ssm_managed_instance_compliance.go:64]:
```suggestion
				Description: "A key-value combination details for the compliance item.",
```

File: aws/table_aws_ssm_managed_instance_compliance.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+				Transform:   transform.FromField("Title"),
+			},
+			{
+				Name:        "resource_id",
+				Description: "An ID for the resource.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "status",
+				Description: "The date when the association was made.",
```
Comment (Subhajit97) [inline aws/table_aws_ssm_managed_instance_compliance.go:44]: Fix description

File: aws-test/tests/aws_ssm_managed_instance_compliance/variables.tf [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+locals {
+  path = "${path.cwd}/compliance.json"
+}
+
+resource "null_resource" "list_compliance" {
+  depends_on = [null_resource.delay]
+  provisioner "local-exec" {
+    command = "aws ssm list-compliance-items --resource-ids ${aws_instance.named_test_resource.id} --resource-types ManagedInstance --output json > ${local.path}"
```
Comment (Subhajit97) [inline aws-test/tests/aws_ssm_managed_instance_compliance/variables.tf:117]: @bigdatasourav can you please provide `--profile` arg?

File: aws/table_aws_ssm_managed_instance_compliance.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+		return nil, err
+	}
+	commonColumnData := commonData.(*awsCommonColumnData)
+
+	akas := []string{"arn:" + commonColumnData.Partition + ":ssm:" + commonColumnData.Region + ":" + commonColumnData.AccountId + ":managed-instance-compliance/" + *data.Id}
+
+	return akas, nil
+}
+
+func getSSMManagedInstanceComplianceTitle(_ context.Context, d *transform.TransformData) (interface{}, error) {
```
Comment (Subhajit97) [inline aws/table_aws_ssm_managed_instance_compliance.go:147]: @bigdatasourav Can we use transform function for `title`, instead of using hydrate function?

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
@bigdatasourav Please see comments, thanks!

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Comment by cbruno10** [comment]
@johnsmyth Now that we have an `aws_ssm_managed_instance` table, does it makes sense to add the compliance information into that table instead of having it as a separate table?

AWS doesn't promote compliance items as first class resources (there is a separate API call, but no defined ARNs), and right now only managed instances are a supported resource type for compliance items (https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_ListComplianceItems.html#API_ListComplianceItems_RequestSyntax):
```
The type of resource from which to get compliance information. Currently, the only supported resource type is ManagedInstance.
```

Before we had a PR that added this information to each association (https://github.com/turbot/steampipe-plugin-aws/pull/359), but then I raised a question on where should that information live, so we tried it as a separate table, but now looking at the example query results in this PR, it seems like it should live in an existing table.

File: aws/table_aws_ssm_managed_instance_compliance.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+)
+
+//// TABLE DEFINITION
+
+func tableAwsSSMManagedInstanceCompliance(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "aws_ssm_managed_instance_compliance",
+		Description: "AWS SSM Managed Instance Compliance",
+		List: &plugin.ListConfig{
+			ParentHydrate: listSsmManagedInstances,
```
Comment (cbruno10) [inline aws/table_aws_ssm_managed_instance_compliance.go:21]: @bigdatasourav I'm a bit worried that for large environments, having this table using parent hydrate will result in a lot of API calls, i.e., hybrid environments that have all of their on-prem instances managed by SSM, along with some EC2 instances too.

Can you please switch this table to require a key, `resource_id`, to reduce the number of calls?

We should then update any examples, and include an example where we join the managed instance results with this table as well to help users.

File: aws/table_aws_ssm_managed_instance_compliance.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+	plugin.Logger(ctx).Trace("getSSMInstanceComplianceAkas")
+	data := h.Item.(*ssm.ComplianceItem)
+
+	commonData, err := getCommonColumns(ctx, d, h)
+	if err != nil {
+		return nil, err
+	}
+	commonColumnData := commonData.(*awsCommonColumnData)
+
+	akas := []string{"arn:" + commonColumnData.Partition + ":ssm:" + commonColumnData.Region + ":" + commonColumnData.AccountId + ":managed-instance-compliance/" + *data.Id}
```
Comment (cbruno10) [inline aws/table_aws_ssm_managed_instance_compliance.go:142]: @bigdatasourav Can you please switch the ARN format to use: `arn:${Partition}:ssm:${Region}:${Account}:managed-instance/${InstanceId}/compliance-item/${ComplianceItemId}:${ComplianceItemType}`? This format is more consistent with AWS' other SSM ARNs, and includes the 3 pieces of information that are unique to each row (compliance ID, compliance type, and resource ID).

**Review by cbruno10 — CHANGES_REQUESTED** [review]
Please see comments, thanks!

File: docs/tables/aws_ssm_managed_instance_compliance.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  id,
+  name,
+  resource_id,
+  status,
+  compliance_type,
+  severity
+from
+  aws_ssm_managed_instance_compliance
+where
+  resource_id = '2a49c5a11ed9d37f';
```
Comment (cbruno10) [inline docs/tables/aws_ssm_managed_instance_compliance.md:24]: @bigdatasourav In the initial example query results, the resource_id columns were EC2 instance IDs, e.g., `i-abc123456def`. Here the resource IDs look a bit different - what type of resource IDs could a managed instance have, and under which circumstances do we get the different types?

File: docs/tables/aws_ssm_managed_instance_compliance.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  status,
+  compliance_type,
+  severity
+from
+  aws_ssm_managed_instance_compliance
+where
+  resource_id = '2a49c5a11ed9d37f'
+  and compliance_type = 'Patch'
+  and status <> 'COMPLIANT';
+```
```
Comment (cbruno10) [inline docs/tables/aws_ssm_managed_instance_compliance.md:61]: @bigdatasourav Can you please also add an example where you're listing managed instances, and then joining those results with this table to list compliance info for all managed instances?

File: docs/tables/aws_ssm_managed_instance_compliance.md (hunk 11 above)
Comment (bigdatasourav) [reply docs/tables/aws_ssm_managed_instance_compliance.md:24]: @cbruno10 I have provided dummy value in the example , resource id will be instace id only.

File: docs/tables/aws_ssm_managed_instance_compliance.md (hunk 12 above)
Comment (bigdatasourav) [reply docs/tables/aws_ssm_managed_instance_compliance.md:61]:
```
### List compliance status for all managed instances

select
  id,
  resource_id,
  status
from
  aws_ssm_managed_instance_compliance
where
  resource_id in (
    select
      instance_id
    from
      aws.aws_ssm_managed_instance
  );

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**